From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: zfcp: Don't create multiple sysfs entries for same WWPN
References: bnc#518291,LTC#54468

Symptom:     The FC transport class creates multiple rport entries in 
             sysfs for the same WWPN.
Problem:     The FC transport class requires that the
             fc_remote_port_add and fc_remote_delete functions are
             always called in this sequence. It was possible that zfcp
             called fc_remote_port_add twice leading to the problem.
Solution:    Make sure to only call fc_remote_port_add once before
             calling fc_remote_port_delete.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/scsi/zfcp_scsi.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/drivers/s390/scsi/zfcp_scsi.c	2009-06-29 14:11:50.000000000 +0200
+++ b/drivers/s390/scsi/zfcp_scsi.c	2009-06-29 14:27:25.000000000 +0200
@@ -525,6 +525,9 @@ static void zfcp_scsi_rport_register(str
 	struct fc_rport_identifiers ids;
 	struct fc_rport *rport;
 
+	if (port->rport)
+		return;
+
 	ids.node_name = port->wwnn;
 	ids.port_name = port->wwpn;
 	ids.port_id = port->d_id;
@@ -548,8 +551,10 @@ static void zfcp_scsi_rport_block(struct
 {
 	struct fc_rport *rport = port->rport;
 
-	if (rport)
+	if (rport) {
 		fc_remote_port_delete(rport);
+		port->rport = NULL;
+	}
 }
 
 void zfcp_scsi_schedule_rport_register(struct zfcp_port *port)
